streamer/pulp/streamer/server.py

```diff
@@ -174,84 +113,183 @@ def _handle_get(self, request):
+            else:
+                raise DownloadFailed()
+        finally:
+            try:
+                downloader.config.finalize()
+            except Exception:
+                # ignored.
+                pass
+
+    def _get_downloader(self, entry):
+        """
+        Get the configured downloader.
+
+        :param entry: A catalog entry.
+        :type  entry: LazyCatalogEntry
+        :return: The configured downloader.
+        :rtype:  nectar.downloaders.base.Downloader
+        :raise: PluginNotFound: when plugin not found.
+        :raise: DoesNotExist: when importer not found.
+        """
+        try:
+            importer, config, model = \
+                repo_controller.get_importer_by_id(entry.importer_id)
+            model.config = config.flatten()
+            downloader = importer.get_downloader_for_db_importer(
+                model, entry.url, working_dir='/tmp')
+            listener = DownloadListener()
+            downloader.event_listener = listener
+            downloader.session = self.session
+            return downloader
+        except (PluginNotFound, DoesNotExist):
+            msg = _('Plugin not-found: referenced by catalog entry for {path}')
+            logger.info(msg.format(path=entry.path))
+            raise
+
+    @staticmethod
+    def _get_unit(entry):
+        """
+        Get the content unit referenced by the catalog entry.
 
-            alt_content_container = content_container.ContentContainer(threaded=False)
-            alt_content_container.download(primary_downloader, [download_request], listener)
+        :param entry: A catalog entry.
+        :type  entry: LazyCatalogEntry 
+        :return: The unit.
+        :raises DoesNotExist: when not found.
+        """
+        try:
+            model = plugin_api.get_unit_model_by_id(entry.unit_type_id)
+            q_set = model.objects.filter(id=entry.unit_id)
+            q_set = q_set.only(*model.unit_key_fields)
+            return q_set.get()
         except DoesNotExist:
-            # A catalog entry is referencing a unit that doesn't exist which is bad.
-            msg = _('The catalog entry for {path} references {unit_type}:{id}, but '
-                    'that unit is not in the database.')
-            logger.error(msg.format(path=catalog_entry.path, unit_type=catalog_entry.unit_type_id,
-                                    id=catalog_entry.unit_id))
-            request.setResponseCode(NOT_FOUND)
-        finally:
-            primary_downloader.config.finalize()
+            msg = _('The catalog entry for {path} references invalid unit: {unit_type}:{id}')
+            logger.error(msg.format(
+                path=entry.path,
+                unit_type=entry.unit_type_id,
+                id=entry.unit_id))
+            raise
+
+    def _set_headers(self, request, report):
+        """
+        Forward response headers to the HTTP request.  This includes adding the
```

Comment:
I think a little more info would make this more clear. Is this the request that a client made to the streamer? Or a request the streamer is about to make to some remote source?

Likewise, what response are we talking about? I assume the ``report`` is representative of the response?

A few words on why these headers need to be copied over would also help.